File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/OperatorCreator.java
package org.aiddl.example.learning_agent.functions;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

import org.aiddl.common.planning.PlanningTerm;
import org.aiddl.core.interfaces.ConfigurableFunction;
import org.aiddl.core.representation.ListTerm;
import org.aiddl.core.representation.SetTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.core.representation.TupleTerm;
import org.aiddl.core.function.FunctionRegistry;

public class OperatorCreator implements ConfigurableFunction {

	int target;
	String name = OperatorCreator.class.getSimpleName();
	boolean verbose = false;

	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.verbose = Term.sym("true").equals(settings.get(Term.sym("verbose")));
	}

	@Override
	public Term apply( Term tree ) {
		List<TupleTerm> O_new = extract(tree, Term.list(), null);
				
		O_new.sort(new CompareTuple());
		SetTerm O_set = Term.set(O_new);
		
		return O_set;
	}
	
	private class CompareTuple implements Comparator<TupleTerm> {
		@Override
		public int compare(TupleTerm o1, TupleTerm o2) {
			return o1.toString().compareTo(o2.toString());
		}
	}
	
	private boolean isCollectionOfSets( SetTerm S ) {
		for ( Term s : S ) {
			if ( !(s instanceof SetTerm) ) {
				return false;
			}
		}
		return true;
	}
	
	private List<TupleTerm> extract( Term tree, ListTerm preCurrent, Term actionName ) {
		List<TupleTerm> O = new ArrayList<TupleTerm>();
		if ( !(tree instanceof ListTerm) ) {
			SetTerm eff = (SetTerm) tree;
			
			if ( this.isCollectionOfSets(eff) ) { 
				SetTerm effects = eff;
				for ( Term eff_inside : effects ) {
					SetTerm actual_eff = (SetTerm) eff_inside;
					
					if ( actual_eff.size() > 0 ) {		
						Collection<Term> preCol = preCurrent.getCollectionCopy();
//						preCol.removeAll(actions.getCollection());
						SetTerm pre = Term.set(preCol);
						
//						Term name = Term.tuple(Term.sym("op"), Term.integer(eff.hashCode()));
//						Term name = Term.tuple(Term.sym("op_" + actual_eff.hashCode()), actual_eff);
						Term name = actionName;

						if ( actionName != null ) {

							Term namePair = Term.keyVal(PlanningTerm.Name, name);
							Term prePair = Term.keyVal(PlanningTerm.Preconditions, pre);
							Term effPair = Term.keyVal(PlanningTerm.Effects, actual_eff);
							
							O.add(Term.tuple(namePair, prePair, effPair));
						} else {
							System.err.println("NOT ADDING...");
						}
					}
				}
			} else {
				if ( eff.size() > 0 ) {
					Collection<Term> preCol = preCurrent.getCollectionCopy();
//					preCol.removeAll(actions.getCollection());
					SetTerm pre = Term.set(preCol);
									
//					Term name = Term.tuple(Term.sym("op"), eff);
//					Term name = Term.tuple(Term.sym("op"), Term.integer(eff.hashCode()));
					Term name = actionName;
					
					if ( actionName != null ) {
						Term namePair = Term.keyVal(PlanningTerm.Name, name);
						Term prePair = Term.keyVal(PlanningTerm.Preconditions, pre);
						Term effPair = Term.keyVal(PlanningTerm.Effects, eff);
											
						O.add(Term.tuple(namePair, prePair, effPair));
					} else {
						System.err.println("NOT ADDING...");
					}
				}
			}
		} else {
			ListTerm decisions = (ListTerm) tree;
			for ( Term decision : decisions ) {
				TupleTerm condition = (TupleTerm) decision.get(0);
				Term subTree = decision.get(1);
				Term preAtom  = condition.get(1);
				Term preValue = condition.get(2);
				
				ListTerm preNew;
				
				if ( preAtom.equals(Term.sym("Action"))) {
					actionName = preValue;
					preNew = preCurrent;
				} else {
					preNew = preCurrent.add(
							Term.keyVal(preAtom, 
										preValue));
				}
		
				O.addAll(extract(subTree, preNew, actionName));
			}
		}
		return O;
	}
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/OperatorCreator.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{CollectionTerm, KeyVal, ListTerm, SetTerm, Sym, Term, Tuple}
import org.aiddl.common.scala.planning.PlanningTerm

import java.util.logging.Level

class OperatorCreator(attributes: ListTerm) extends Function with Verbose {
  this.logConfig(level=Level.WARNING)

  override def apply(tree: Term): Term = {
    this.logger.info("Starting operator extraction")
    val oNew = this
      .extract(tree, ListTerm.empty, null)
      .sortWith((a, b) => a.toString.compareTo(b.toString) < 0)
    this.logger.info(s"Found ${oNew.size} operators")
    SetTerm(oNew.toSet)
  }

  private def isCollectionOfSets(s: CollectionTerm): Boolean =
    s.forall( t => t.isInstanceOf[SetTerm] )

  private def extract(tree: Term, preCurrent: ListTerm, actionName: Term): List[Term] = {
    var O: List[Term] = Nil
    if (!tree.isInstanceOf[ListTerm]) {
      val eff = tree.asList

      if (this.isCollectionOfSets(eff)) {
        for (eff_inside <- eff) {
          val actual_eff = eff_inside.asSet
          if (actual_eff.nonEmpty) {
            val preCol = preCurrent.asSet
            if (actionName != null) {
              val namePair = KeyVal(PlanningTerm.Name, actionName)
              val prePair = KeyVal(PlanningTerm.Preconditions, preCol)
              val effPair = KeyVal(PlanningTerm.Effects, actual_eff(Sym("class")))
              val o = Tuple(namePair, prePair, effPair)
              O = o :: O
            }
          }
        }
      } else if (eff.size > 0) {
        val name = actionName
        if (actionName != null) {
          val namePair = KeyVal(PlanningTerm.Name, name)
          val prePair = KeyVal(PlanningTerm.Preconditions, preCurrent.asSet)
          val effPair = KeyVal(PlanningTerm.Effects, eff(Sym("class")))
          val o = Tuple(namePair, prePair, effPair)
          O = o :: O
        }
      }
    } else {
      val decisions = tree.asList
      for (decision <- decisions) {
        var usedActionName = actionName
        val condition = decision.asTup(0)
        val subTree = decision(1)
        val preAtom = attributes(condition(1).intoInt)
        val preValue = condition(2)
        val preNew: ListTerm = if (preAtom.equals(Sym("Action"))) {
            usedActionName = preValue
            preCurrent
          } else {
            preCurrent.add(KeyVal(preAtom, preValue))
          }
          O = O.prependedAll(extract(subTree, preNew, usedActionName))
        }
      }
    O
  }
}
